keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/NotificationEventListener.kt

```diff
@@ -0,0 +1,87 @@
+import com.netflix.spinnaker.keel.events.UnpinnedNotification
+import com.netflix.spinnaker.keel.persistence.KeelRepository
+import com.netflix.spinnaker.keel.slack.handlers.SlackNotificationHandler
+import org.slf4j.LoggerFactory
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.context.event.EventListener
+import org.springframework.stereotype.Component
+import java.time.Clock
+
+/**
+ * Responsible to listen to notification events, and fetch the information needed
+ * for sending a notification, based on NotificationType.
+ */
+@Component
+class NotificationEventListener(
+  private val repository: KeelRepository,
+  private val publisher: ApplicationEventPublisher,
+  private val clock: Clock,
+  private val handlers: List<SlackNotificationHandler<*>>
+) {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  @EventListener(PinnedNotification::class)
+  fun onPinnedNotification(notification: PinnedNotification) {
+    with(notification) {
+      val envNotifications = repository.environmentNotifications(config.name, pin.targetEnvironment)
+      val deliveryArtifact = pin.reference.let {
+        repository.getArtifact(config.name, it)
+      }
+      val pinnedArtifact = repository.getArtifactVersion(deliveryArtifact, pin.version, null)
+      val currentArtifact = repository.getArtifactVersionByPromotionStatus(config, pin.targetEnvironment, deliveryArtifact, PromotionStatus.CURRENT)
+
+      //handlers.supporting<SlackPinnedNotification>(type)
+
+      envNotifications?.forEach { notificationConfig ->
+        if (notificationConfig.type == NotificationType.slack) {
+          val handler = handlers.find { hand ->
+            hand.type == notification.type
+          } as? SlackNotificationHandler<SlackPinnedNotification>
+          handler?.let {
+            it.constructMessage(SlackPinnedNotification(
+              channel = notificationConfig.address,
+              pin = pin,
+              currentArtifact = currentArtifact,
+              pinnedArtifact = pinnedArtifact,
+              application = config.application,
+              time = clock.instant()
+            ))
+          }
+        }
+      }
+    }
+  }
+
+  @EventListener(UnpinnedNotification::class)
+  fun onUnpinnedNotification(notification: UnpinnedNotification) {
+    with(notification) {
+      val envNotifications = repository.environmentNotifications(config.name, targetEnvironment)
+
+      if (pinnedEnvironment != null) {
+        val latestApprovedArtifactVersion = repository.latestVersionApprovedIn(config, pinnedEnvironment!!.artifact, targetEnvironment)
+        val latestArtifact = latestApprovedArtifactVersion?.let { repository.getArtifactVersion(pinnedEnvironment!!.artifact, it, null) }
+
+        envNotifications?.forEach { notificationConfig ->
+          if (notificationConfig.type == NotificationType.slack) {
+            publisher.publishEvent(SlackUnpinnedNotification(
```

Comment:
What's the purpose of this event listener raising this other event here?

Reply:
oh this is the old implementation -- it won't send an event, rather it will call the right handler. unpinning is still in TODO :)

Reply:
What do you mean it won't send an event? Isn't that what the code is doing here? You mean you'll replace that with new code in a separate PR? If that's the case, it might be good to just remove this code from here, because it's confusing IMHO.

Reply:
Never mind, I see the new code now!